### ISSUE TITLE
Flux contents ignored by l10nmgr extension in TYPO3 v12

### ISSUE THREAD

**Issue by TopSolid** [FluidTYPO3/flux#2163]
**I have checked that the bug exists in the `dev-development` branch**
Yes
If you're about to answer "no" here, it is your responsibility to check `dev-development` **before** opening a bug report!

**I have checked that there are no already open issues or recently closed issues about this bug**
Yes

If you're about to answer "no" here, it is your responsibility to first check the open and recently closed issues for potential duplicaates.

**Describe the bug**
I use l10nmgr in v12 and it ignores all flux contents when exporting in XML.
Despite latest versions have been used:

TYPO3 12.4.15
fluidtypo3/flux dev-development 907b7ab
l10nmgr EA 12 version from 21/06 release

With v11, a patch was possible https://github.com/FluidTYPO3/flux/issues/1805 where a function "patchTceformsWrapper()" was missing. Since v12, it seems that it is not necessary anymore: https://github.com/FluidTYPO3/flux/blob/development/Classes/Builder/FlexFormBuilder.php#L123

Does it come from flux or l10nmgr? not easy to find... Can anyone here help?

**Comment by NamelessCoder** [comment]
As far as I am aware this method is not necessary at all on TYPO3v12 and is only partially necessary on TYPO3v11 and below, which triggers some edge case errors with things like the suggest XHR receiver without the TCEforms wrapper.

It should _not_ be necessary at all with the most recent version of EXT:l10nmgr which has its own patchTceFormsWrapper method (apparently a 1:1 copy of the method from Flux, added in v11.0 of EXT:l10nmgr): https://github.com/CodersCare/l10nmgr/blob/3ee71720066bafdf2c5abd736c1f97c6f3ae61ec/Classes/Model/Tools/Tools.php#L551.

In short: even if Flux doesn't apply this wrapper on TYPO3v12, l10nmgr will apply it, as long as you use v11.0 or higher of l10nmgr. I would certainly expect this patch to also be in l10nmgr EA.

If that doesn't solve your problem there might be another cause for it, but I can't help you track it down since I don't have EA access for l10nmgr. You'd have to bring this problem to the l10nmgr team.

**Comment by Bunnyfield** [comment]
@TopSolid-DaA @NamelessCoder I just dug a bit deeper into that part of the code to find out the difference that makes usual FlexForms work, while the Flux generated XML still seems to break the L10nmgr export.

It seems the actual problem is caused by a core bug, which is again caused by the replacement of

\TYPO3\CMS\Core\Configuration\FlexForm\FlexFormTools::getArrayValueByPath

with

\TYPO3\CMS\Core\Utility\ArrayUtility::getValueByPath

While the deprecation message says: "is deprecated since v11 and will be removed in v12. Use ArrayUtility::getValueByPath() instead.'," the new method is missing an essential part, that was still available with the original FlexFormTools.

```
        if (is_array($array) && !empty($pathArray)) {
            $key = array_shift($pathArray);
            if (isset($array[$key])) {
                if (empty($pathArray)) {
                    return $array[$key];
                }
                return $this->getArrayValueByPath($pathArray, $array[$key]);
            }
            return null;
        }

```

This specific part makes sure, that the method gets called recursively, as soon as the value is not a string but an array itself.
So the returned value will always be a string, while the new method just stops and returns the array without going any deeper into the structure.

```
        foreach ($path as $segment) {
            if (is_array($value) && array_key_exists($segment, $value)) {
                // Replace current value with child
                $value = $value[$segment];
            } else {
                // Throw specific exception if there is no such path
                throw new MissingArrayPathException('Segment ' . $segment . ' of path ' . implode($delimiter, $path) . ' does not exist in array', 1341397869);
            }
        }
        return $value;
```

So it's not enough to migrate to the new method as the deprecation message said, instead we need to reintroduce the recursion to make sure even nested array structure paths used by Flux-based FlexForms will be returned properly.

Since other FlexForms did not use deeply nested structures, they were still able to be exported properly.

**Comment by Bunnyfield** [comment]
@TopSolid-DaA After sleeping on it, my current approach would be to revive the missing functionality within the L10nmgr call. That way the core doesn't need to be changed and it will be fixed for other translation scenarios with deeply nested FlexForms. I think the patch will be ready tonight.

**Comment by Bunnyfield** [comment]
@TopSolid-DaA See https://github.com/CodersCare/l10nmgr_EA/issues/96#issuecomment-2457901729
